### PR TITLE
8319197: Exclude hb-subset and hb-style from compilation

### PR DESCRIPTION
**Pull request by djelinski**
hb-subset and hb-style APIs are not used and not exported by libfontmanger. We can cut the compilation time by not compiling the unused files.

The added exclusions reduce the build time by ~1 minute (~8%) on my machine. This is with harfbuzz 8, with harfbuzz 7 the improvement was ~30 seconds.

Client libs tests continue to pass.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8319197](https://bugs.openjdk.org/browse/JDK-8319197): Exclude hb-subset and hb-style from compilation (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16440/head:pull/16440` \
`$ git checkout pull/16440`

Update a local copy of the PR: \
`$ git checkout pull/16440` \
`$ git pull https://git.openjdk.org/jdk.git pull/16440/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16440`

View PR using the GUI difftool: \
`$ git pr show -t 16440`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16440.diff">https://git.openjdk.org/jdk/pull/16440.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16440#issuecomment-1788642843)

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]
This looks good to me from a build point of view, but needs someone from client to approve as well.

/reviewers 2

**Review by aivanov-jdk — APPROVED** [review]
Looks good to me.

You may want to wait for Phil's approval though.

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk
Comment (prrace) [inline make/modules/java.desktop/lib/Awt2dLibraries.gmk:499]: I'm quite confused from where you got gsubgpos-context.cc. I don't see any file of that name.
Also I just realised that since the Nov 2020 import of freetype it has a bunch of handly ifdefs that
you can see in hb-config.h
And so probably a better way of doing this, even if it means one of these files (the instance-solver one) doesn't seem to get the ifdef treatment is to use the specific defines passed to the build.
Then I'd expect harfbuzz to apply the same ifdef even if these files got refactored.
I further noticed that the same 2020 update adds HAVE_FREETYPE and hb-ft.cc isn't compiled unless you actually set that. Which we don't
So this whole exclude can go away and be replaced by a few -D defines.
You could probably go overboard and experiment with all sorts of combinations from hb-config.h but I recommend you just apply the minimal set that makes your build time happier.
hb-style.cc isn't one I'd have realised we don't currently need .. and for many of them it isn't clear to me without research what the consequences would be of excluding it. Without that support would something non-optimal happen in some cases - you can't trust the tests we have to come near proving it.
I know we don't subset, and that's the only one I consider "safe" to use here without actual research.

Comment (prrace) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:499]: These two
#define HB_NO_SUBSET_LAYOUT
#define HB_NO_SUBSET_CFF

Comment (djelinski) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:499]: You can find the file here:
https://github.com/openjdk/jdk/blob/f262f06c97b9ea94cd6119b3a8beb16bf804d083/src/java.desktop/share/native/libharfbuzz/graph/gsubgpos-context.cc

Hb-subset files are not guarded by ifdefs; HB_NO_SUBSET_LAYOUT only excludes some functions, but not others. I'll check the impact of that define when I'm back in office.

I checked the list of HB APIs [here](https://harfbuzz.github.io/reference-manual.html) ; hb-subset and hb-style belong to non-core API sets, so I decided to check if hb-style was needed. It was not.

We can safely stop compiling the files listed because:
- none of the harfbuzz functions are exported, and they are not accessible outside of libfontmanager.
- if we excluded any file that was actually used, linker would refuse to link libfontmanager

Given that we only copy an arbitrary subset of harfbuzz files anyway (see UPDATING.txt), would it make sense to remove these files instead?

Comment (djelinski) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:499]: FWIW, I compiled the code without this PR, but with `HARFBUZZ_CFLAGS += -DHB_NO_SUBSET_LAYOUT -DHB_NO_SUBSET_CFF` instead, and checked `make LOG=profile` output. Results:
- without this change, compiling `hb-subset.cc` took 56 seconds, and `hb-subset-plan.cc` took 33 seconds
- with this change, compiling `hb-subset.cc` took 33 seconds, and `hb-subset-plan.cc` took 22 seconds

It's a nice improvement, but not compiling these files at all is still much better.

Comment (magicus) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:499]: @djelinski Just curious; what would the effect be to both include this change and setting the NO_* defines?

If all the references to these defines are made in the excluded files then the only reason for doing that would be some kind of information to subsequent readers of the code, but they might also be checked elsewhere, and thus give an additional speedup.

Comment (djelinski) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:499]: HB_NO_SUBSET_CFF is indeed referenced in 2 header files, which are in turn included somewhere else.
I added the defines on top of the excludes, but this resulted in no measurable build time difference.

Comment (djelinski) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:499]: Out of curiosity, I tried building with [HB_LEAN](https://github.com/openjdk/jdk/blob/2d4bbf478745e62584420bfdef5a4948edac54ad/src/java.desktop/share/native/libharfbuzz/hb-config.hh#L57), which is a shortcut for "no optional features"; I got a couple of new warnings, and one error (`hb_font_funcs_set_glyph_v_kerning_func` is deprecated since 2.0 and not compiled with HB_LEAN; fortunately its use in hb-jdk-font.cc can be simply commented out). After resolving these problems, fontmanager compiled successfully - but again, no measurable time difference.

**Comment by prrace** [comment]
So if I have this right,
(1) excluding several unused files from the build saves 60 seconds.
(2) Using the two specific #defines instead saves about 34 seconds.
(3) The two specific #defines on top of the excludes from (1) made no difference
(4) HB_LEAN had no impact either .. I presume you mean  when used on top of (1) as well, which suggests that most of what's in there is cheap to compile.

The build exclusion does however include another couple of files, at least one of which is unrelated, not sure if you ever said
how much that saved, but that presumably accounts for some part of 60 sec vs 34 secs

**Comment by djelinski** [comment]
(1) - (4) - that's correct.

My tests were hardly scientific; most of the time I only deleted and recompiled libfontmanager using 8 parallel jobs; the time taken by the recompilation was in the 50-75 seconds range both with and without HB_LEAN.
Without the exclusions, the rebuild of libfontmanager takes ~2 minutes.

File build times from a sample run, taken from build-profile.log:
| file | time without any changes | time with -DHB_NO_SUBSET_LAYOUT -DHB_NO_SUBSET_CFF |
| -- | -- | -- |
hb-subset-cff-common.cc | 7.46s | 2.68s
hb-subset-cff1.cc | 14.06s | 2.47s
hb-subset-cff2.cc | 13.14s | 2.82s
hb-subset-input.cc | 5.70 s | 4.36s
hb-subset-plan.cc | 33.62s | 22.27s
hb-subset.cc | 56.15s | 33.33s
hb-subset-instancer-solver.cc | 4.81s | 2.81s
gsubgpos-context.cc | 11.66s | 7.48s
hb-style.cc | 9.49s | 6.67s

**Review by prrace — APPROVED** [review]
OK, let's just go with the exclusion at least for now.
If it becomes a problem on a subsequent upgrade it may need to be changed.
We can also explore other options later.

**Comment by djelinski** [comment]
Thanks!

/integrate

**Pull request closed without merging**
